morphic.js

```diff
@@ -7818,10 +7820,47 @@ MenuMorph.prototype.select = function (aMenuItem) {
     this.selection = aMenuItem;
 };
 
+MenuMorph.prototype.mouseLeave = function () {
+    if (this.parent instanceof MenuMorph) {
+        // this is a submenu so we vanish when lose focus
+        var mouses = this.world.hand.allMorphsAtPointer();
+        if (contains(mouses, this.parent)) {
+            var items = mouses.filter(function (morph)
```

Comment:
Local variable declarations – using `var` – are scoped to the function, so it usually makes more sense to declare the variable at the top of the function:

```js
function foo(arg) {
  console.log(x); // the value undefined
  console.log(y); // throws a reference error

  if (arg === 3) {
    var x = 123;
  } else {
    var x = 15;
  }

  console.log(x); // 123 or 15
}
```

See how `x` had the value `undefined` but when we tried to get the value of `y`, we got a reference error? This becomes a real problem in a case like this:

```js
function foo(x) {
  return function() {
    var y = x + 2

    // later in the logic..

    if (Math.random() > 0.5) {
      var x = 15 * 7 * 1.5
      y += x
    }

    return y
  }
}
```

So if we call `foo(3)`, we get a function that would *theoretically* return either `5` (`y = x + 2`) or `162.5` (`y` (5) `+= 15 * 7 * 1.5`), but we *actually* get `NaN`! That's because our later declaration of `var x = 15 * 7 * 1.5` overshadowed the earlier declaration of `x` (as an argument to `foo`), so when we did `var y = x + 2` we actually did `var y = undefined + 2`, which is `NaN`. (Any later math we might do to `NaN` also gets us `NaN`; that's just a property of `NaN`.)

So if we move the variable declarations to the top, we end up getting this:

```js
function foo(x) {
  return function() {
    var x // <--- 'var x' moved here
    var y = x + 2

    // later in the logic..

    if (Math.random() > 0.5) {
      x = 15 * 7 * 1.5 // <--- no var
      y += x
    }

    return y
  }
}
```

..and then it's clear that `x` is `undefined` when we add it in `y = x + 2`, so we've figured out the problem.

Convoluted examples aside, hopefully this makes at least a little bit of sense :)

IMO this is one of the biggest problems with JavaScript, but lately it's gotten better since you can use `let` and `const`, which are scoped to the *block* they're in:

```js
function foo(arg) {
  console.log(x); // throws a reference error

  let x;
  if (arg === 3) {
    x = 123;
  } else {
    x = 15;
  }

  console.log(x); // 123 or 15
}
```